library/Vanilla/TokenSigningTrait.php

```diff
@@ -0,0 +1,186 @@
+<?php
+/**
+ * @author Chris Chabilall chris.c@vanillaforums.com
+ * @copyright 2009-2018 Vanilla Forums Inc.
+ * @license http://www.opensource.org/licenses/gpl-2.0.php GNU GPL v2
+ */
+
+namespace Vanilla;
+
+/**
+ * Methods to be used for token generation and signing.
+ */
+trait TokenSigningTrait {
+    /** @var string $secret The secret used to sign the token. */
+    private $secret;
+
+    /** @var string Used to determine what type of token is generated. */
+    protected $tokenIdentifier;
+
+    /**
+     * Get the secret.
+     *
+     * @return mixed Returns the secret.
+     */
+    public function getSecret() {
+        return $this->secret;
+    }
+
+    /**
+     * Set the secret.
+     *
+     * @param mixed $secret
+     */
+    public function setSecret($secret) {
+        $this->secret = $secret;
+        return $this;
+    }
+
+    /**
+     * Base64 Encode a string, but make it suitable to be passed in a url.
+     *
+     * @param string $str The string to encode.
+     * @return string The encoded string.
+     */
+    private static function base64urlEncode($str) {
+        return trim(strtr(base64_encode($str), '+/', '-_'), '=');
+    }
+
+    /**
+     * Decode a string that was encoded using base64UrlEncode().
+     *
+     * @param string $str The encoded string.
+     * @return string The decoded string.
+     */
+    private static function base64urlDecode($str) {
+        return base64_decode(strtr($str, '-_', '+/'));
+    }
+
+    /**
+     * Generate a random token.
+     *
+     * This token will go into the database, but must be signed before being given to the client.
+     *
+     * @return string Returns a 32 character token string.
+     */
+    public function randomToken() {
+        $token = self::base64urlEncode(openssl_random_pseudo_bytes(24));
+        return $token;
+    }
+
+    /**
+     * Verify a token's signature.
+     *
+     * @param string $accessToken The full access token.
+     * @param bool $throw Whether or not to throw an exception on a verification error.
+     * @return bool Returns **true** if the token's expiry date and signature is valid or **false** otherwise.
+     * @throws \Exception Throws an exception if the token is invalid and {@link $throw} is **true**.
+     */
+    public function verifyTokenSignature($accessToken, $throw = false) {
+        $parts = explode('.', $accessToken);
+
+        if (empty($accessToken)) {
+            return $this->tokenError('Missing '.$this->tokenIdentifier, 401, $throw);
+        }
+
+        if (count($parts) !== 4) {
+            return $this->tokenError($this->tokenIdentifier.' missing parts.', 401, $throw);
```

Comment:
This error message could potentially read "nonce missing parts.", which doesn't align with the other messages in this method all starting with a capital letter. You can prefix the message with "Your", like similar messages, below.